check/format-incremental

```diff
@@ -35,77 +36,66 @@ cd "$(git rev-parse --show-toplevel)"
+        else
+            echo -e "\033[31mNo default revision found to compare against. Argument #1 must be what to diff against (e.g. 'origin/master' or 'HEAD~1').\033[0m" >&2
+            exit 1
+        fi
+    fi
+    base="$(git merge-base ${rev} HEAD)"
+    if [ "$(git rev-parse ${rev})" == "${base}" ]; then
+        echo -e "Comparing against revision '${rev}'." >&2
     else
-        echo -e "\033[31mNo default revision found to compare against. Argument #1 must be what to diff against (e.g. 'origin/master' or 'HEAD~1').\033[0m" >&2
-        exit 1
+        echo -e "Comparing against revision '${rev}' (merge base ${base})." >&2
+        rev="${base}"
     fi
-fi
-base="$(git merge-base ${rev} HEAD)"
-if [ "$(git rev-parse ${rev})" == "${base}" ]; then
-    echo -e "Comparing against revision '${rev}'." >&2
+
+    # Get the _test version of changed python files.
+    format_files=$(git diff --name-only ${rev} -- | grep "\.py$" | grep -v "_pb2\.py$")
 else
-    echo -e "Comparing against revision '${rev}' (merge base ${base})." >&2
-    rev="${base}"
+    echo -e "Formatting all python files." >&2
+    format_files=$(find . -name "*.py" | grep -v "_pb2\.py$")
 fi
 
-# Get the _test version of changed python files.
-needed_changes=0
-changed_files=$(git diff --name-only ${rev} -- | grep "\.py$" | grep -v "_pb2\.py$")
-esc=$(printf '\033')
-for changed_file in ${changed_files}; do
-    # Extract changed line ranges from diff output.
-    changed_line_ranges=$( \
-        git diff --unified=0 "${rev}" -- "${changed_file}" \
-        | perl -ne 'chomp(); if (/@@ -\d+(,\d+)? \+(\d+)(,)?(\d+)? @@/) {$end=$2+($4 or 1)-1; print "--lines=$2-$end "}' \
-    )
-
-    if [[ "${changed_line_ranges}" != "--lines=0-0 " ]]; then
-        # Do the formatting.
-        results=$(yapf --style=google --diff "${changed_file}" ${changed_line_ranges})
-
-        # Print colorized error messages.
-        if [ ! -z "${results}" ]; then
-            needed_changes=1
-            if (( only_print == 0 )); then
-                $(yapf --style=google --in-place "${changed_file}" ${changed_line_ranges})
-            else
-                echo -e "\n\033[31mChanges in ${changed_file} require formatting:\033[0m\n${results}" \
-                    | sed "s/^\(+ .*\)$/${esc}[32m\\1${esc}[0m/" \
-                    | sed "s/^\(- .*\)$/${esc}[31m\\1${esc}[0m/"
-            fi
-        fi
-     fi
+any_files=0
+for format_file in ${format_files}; do
+  any_files=1
 done
 
-if (( needed_changes == 0 )); then
-    echo -e "\033[32mNo formatting needed on changed lines\033[0m."
-elif (( only_print == 1 )); then
-    echo -e "\033[31mSome formatting needed on changed lines\033[0m."
-    exit 1
+if (( any_files == 0 )); then
+    echo -e "\033[32mNo files to format\033[0m."
 else
-    echo -e "\033[33mReformatted changed lines\033[0m."
+    args=()
+    if (( only_print == 1 )); then
+        args=("--check" "--diff")
+    fi
+    black --color "${args[@]}" ${format_files}
```

Comment:
Shouldn't we run twice with --fast to work around https://github.com/psf/black/issues/1629?

Reply:
It does seem like it...